### PR TITLE
Make `get_fptr_shape` public and a bug fix

### PR DESCRIPTION
**Pull request by pchakraborty**
## Types of change(s)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (affects only documentation or cleanup)
- [ ] Refactor (no functional changes, no api changes)

## Checklist
- [ ] Tested this change with a run of GEOSgcm
- [ ] Ran the Unit Tests (`make tests`)

## Description
- Excluding negative values of gridToFieldMap
- Making `mapl3g_FieldCondensedArray::get_fptr_shape` public

### CODE REVIEW

File: field_utils/FieldCondensedArray_private.F90 [hunk 1]
```diff
@@ -26,9 +26,9 @@ function get_fptr_shape_private(gridToFieldMap, localElementCount, has_vertical,
       
       vert_dim = 0
       vert_size = 1
-      
+
       rank = size(localElementCount)
-      grid_dims = pack(gridToFieldMap, gridToFieldMap /= 0)
+      grid_dims = pack(gridToFieldMap, gridToFieldMap > 0)
```
Comment (tclune) [inline field_utils/FieldCondensedArray_private.F90:31]: Can you elaborate on the circumstances where elements of gridToFieldMap are negative?   My memory is that the documentation says "0" or the index of ...

File: field_utils/FieldCondensedArray.F90 [hunk 2]
```diff
@@ -11,6 +11,7 @@ module mapl3g_FieldCondensedArray
    implicit none
    private
    public :: assign_fptr_condensed_array
+   public :: get_fptr_shape
```
Comment (tclune) [inline field_utils/FieldCondensedArray.F90:14]: The intent is that the vertical regridder use the higher level procedure that simply returns a pointer to the correctly shaped 3D array.

We will need to understand why this does not work in your use case.

File: field_utils/FieldCondensedArray_private.F90 (hunk 1 above)
Comment (pchakraborty) [reply field_utils/FieldCondensedArray_private.F90:31]: For the field below,  `gridToFieldMap` is [ 1, 2, -1]
```
Field Print Starts ====>
 Field base status = Ready
   Name = 'Field034'
 Base name    = Field034
 Status: Base = Ready,  object = Ready
 Proxy        = no
 Root Info (Attributes):
{
  "MAPL": {
    "internal": {
      "long_name": "unknown",
      "standard_name": "E",
      "ungridded_dims": {
        "num_ungridded_dimensions": 0
      },
      "units": "m",
      "vertical_dim": {
        "vloc": "VERTICAL_DIM_CENTER"
      },
      "vertical_grid": {
        "num_levels": 3
      }
    }
  }
} Field status =            4
 gridrank =            2
--- ESMCI::Array::print start ---
 Base name    = Field034
 Status: Base = Ready,  object = Ready
 Proxy        = no
 Root Info (Attributes):
{}Array typekind/rank: Real*4 / 3
~ lower class' values ~
DistGrid dimCount = 2
deCount = 1
localDeCount = 1
~ local data in LocalArray for DE 0 ~
LocalArrayPrint: Array at address 0x23fe6f0:
            rank = 3, typekind = 5, base_addr = 0x14aab7580100
            dim[0] = 12  dim[1] = 13  dim[2] = 3
  Real, *4, Dim 3, Data values:
( 1, 1, 1) =  17
  skipping to end ...
(12,13, 3) =  17
dim 1: [0]: [1 [1 [1, 12] 12] 12]
dim 2: [1]: [1 [1 [1, 13] 13] 13]
dim 3: undistLBound[0]=1            undistUBound[0]=3
--- ESMCI::Array::print end ---
 arrayrank =            3
 gridToFieldMap ungriddedLBound ungriddedUBound totalLWidth totalUWidth
           1           1           3                0           0
           2          -1          -1                0           0
          -1          -1          -1               -1          -1
          -1          -1          -1               -1          -1
          -1          -1          -1               -1          -1
          -1          -1          -1               -1          -1
          -1          -1          -1               -1          -1
 Field Print Ends   ====>
```

File: field_utils/FieldCondensedArray.F90 (hunk 2 above)
Comment (pchakraborty) [reply field_utils/FieldCondensedArray.F90:14]: This doesn't need to be public when using the higher level function `assign_fptr_condensed_array`. I'll make the change.

File: field_utils/FieldCondensedArray_private.F90 (hunk 1 above)
Comment (tclune) [reply field_utils/FieldCondensedArray_private.F90:31]: `gridToFieldMap` should only have 2 elements in for this field:

List with number of elements equal to the grid’s dimCount.

This suggests the bug is elsewhere when allocating the array.

Comment (tclune) [reply field_utils/FieldCondensedArray_private.F90:31]: I think the problem is in the higher level routine.   `dimCount` was used to allocate the array but it should be `geomDimCount`.    Not that the man page makes the distinction particularly clear.

Comment (tclune) [reply field_utils/FieldCondensedArray_private.F90:31]: @darianboggs - please determine why the existing tests did not cover this case (and then help @pchakraborty to remedy the tests as part of this PR.)

Comment (pchakraborty) [reply field_utils/FieldCondensedArray_private.F90]: > I think the problem is in the higher level routine. `dimCount` was used to allocate the array but it should be `geomDimCount`. Not that the man page makes the distinction particularly clear.

I have pushed this change.

**Review by tclune — APPROVED** [review]
I would still like to see a unit test that guards against this issue.   This is the 2nd bug that has been found in this interface.  First was that the gridToFieldMap was not even being allocated, and the unit tests only covered the lower level procedures.  (@darianboggs)

**Comment by pchakraborty** [comment]
> I would still like to see a unit test that guards against this issue. This is the 2nd bug that has been found in this interface. First was that the gridToFieldMap was not even being allocated, and the unit tests only covered the lower level procedures. (@darianboggs)

Created issue #3127

**Review by darianboggs — COMMENTED** [review]
It looks good.

I'm not sure what the use case for `get_fptr_shape` being public is, but I'm not opposed. The intention was that `assign_fptr_condensed_array` would get the fptr without the user needing to know the shape. `get_fptr_shape` is used to get the shape according to the Condensed Array convention. `assign_fptr_condensed_array` uses `assign_fptr` in `FieldPointerUtilities.F90` to get the fptr. Once you have the fptr,  you can get the shape with the Fortran `shape` function.

In other words, `FieldCondensedArray.F90` depends on `FieldPointerUtilities.F90`, so you do not need to use both to get the fptr, and you don't need `get_fptr_shape` to get the shape of the fptr outside of `FieldCondensedArray.F90`

We could move all of `FieldCondensedArray.F90` to `FieldPointerUtilities.F90` with `assign_fptr_condensed_array` public and `get_fptr_shape` private. 

`FieldCondensedArray_private.F90` is internal to `field_utils` with public methods for the sake of unit testing.

Let me know if there is an unmet need.

**Comment by tclune** [comment]
> I'm not sure what the use case for `get_fptr_shape` being public is, but I'm not opposed.

I think you are out of date.  That was removed.

**Pull request merged**